File: src/main/java/net/gtaun/shoebill/object/Label.java
/**
 * Copyright (C) 2011-2014 MK124
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.object;

import java.util.Collection;

import net.gtaun.shoebill.SampObjectManager;
import net.gtaun.shoebill.data.Color;
import net.gtaun.shoebill.data.Location;
import net.gtaun.shoebill.data.Vector3D;
import net.gtaun.shoebill.exception.CreationFailedException;

/**
 *
 *
 * @author MK124
 */
public interface Label extends Destroyable, Proxyable<Label>
{
    /**
     * Gets a Label by its Id
     * @param id The Id
     * @return The found Label
     */
	public static Label get(int id)
	{
		return SampObjectManager.get().getLabel(id);
	}

	public static Collection<Label> get()
	{
		return SampObjectManager.get().getLabels();
	}

    /**
     * Create a Label with params. If the Creation fails, it will throw a CreationFailedException.
     *
     * @param text The displayed Text.
     * @param color The color of the Label.
     * @param loc Location where the Label should be.
     * @param drawDistance The Drawdistance of the Label.
     * @param testLOS If the Label can be seen through objects.
     * @return The created Label.
     */
	public static Label create(String text, Color color, Location loc, float drawDistance, boolean testLOS) throws CreationFailedException
	{
		return SampObjectManager.get().createLabel(text, color, loc, drawDistance, testLOS);
	}

    /**
     * Create a Label with params. If the Creation fails, it will throw a CreationFailedException.
     *
     * @param text The displayed Text.
     * @param color The color of the Label.
     * @param x X-Pos where the Label should be.
     * @param y Y-Pos where the Label should be.
     * @param z Z-Pos where the Label should be.
     * @param worldId Worldid where the Label should be.
     * @param drawDistance The Drawdistance of the Label.
     * @param testLOS If the Label can be seen through objects.
     * @return The created Label.
     */
	public static Label create(String text, Color color, float x, float y, float z, int worldId, float drawDistance, boolean testLOS) throws CreationFailedException
	{
		return SampObjectManager.get().createLabel(text, color, x, y, z, worldId, drawDistance, testLOS);
	}

    /**
     * Create a Label with params. If the Creation fails, it will throw a CreationFailedException.
     *
     * @param text The displayed Text.
     * @param color The color of the Label.
     * @param pos Vector3D-Pos where the Label should be.
     * @param worldId Worldid where the Label should be.
     * @param drawDistance The Drawdistance of the Label.
     * @param testLOS If the Label can be seen through objects.
     * @return The created Label.
     */
	public static Label create(String text, Color color, Vector3D pos, int worldId, float drawDistance, boolean testLOS) throws CreationFailedException
	{
		return SampObjectManager.get().createLabel(text, color, pos, worldId, drawDistance, testLOS);
	}


	public static final int INVALID_ID = 0xFFFF;


    /**
     * Gets the Id of the Label
     * @return The Id of the Label
     */
	int getId();

    /**
     * Gets the Text of the Label
     * @return The Text of the Label
     */
	String getText();

    /**
     * Gets the Color of the Dialog
     * @return The Color of the Dialog
     */
	Color getColor();

    /**
     * Gets the draw distance of the Label.
     * @return The Draw Distance
     */
	float getDrawDistance();

    /**
     * Gets the Location of the Label
     * @return The Location
     */
	Location getLocation();

    /**
     * Gets the attached Player. If there is no, this Function will return null.
     * @return The attached Player
     */
	Player getAttachedPlayer();

    /**
     * Gets the attached Vehicle. If there is no, this Function will return null.
     * @return The attached Vehicle
     */
	Vehicle getAttachedVehicle();

    /**
     * Attaches the Label to a Player
     * @param player The Player
     * @param x X-Offset
     * @param y Y-Offset
     * @param z Z-Offset
     */
	void attach(Player player, float x, float y, float z);

    /**
     * Attaches the Label to a Player
     * @param player The Player
     * @param offset The Location
     */
	void attach(Player player, Vector3D offset);

    /**
     * Attaches the Label to a Vehicle
     * @param vehicle The Vehicle
     * @param x X-Offset
     * @param y Y-Offset
     * @param z Z-Offset
     */
	void attach(Vehicle vehicle, float x, float y, float z);

    /**
     * Attaches the Label to a Vehicle
     * @param vehicle The Vehicle
     * @param offset The Location
     */
	void attach(Vehicle vehicle, Vector3D offset);

    /**
     * Updates the Text and Color of the Label
     * @param color New Color
     * @param text New Text
     */
	void update(Color color, String text);
}

File: src/main/java/net/gtaun/shoebill/object/Label.kt
/**
 * Copyright (C) 2011-2014 MK124

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.`object`

import net.gtaun.shoebill.SampObjectManager
import net.gtaun.shoebill.constant.Collectable
import net.gtaun.shoebill.constant.Findable
import net.gtaun.shoebill.data.Color
import net.gtaun.shoebill.data.Location
import net.gtaun.shoebill.data.Vector3D
import net.gtaun.shoebill.exception.CreationFailedException

/**
 * This class wraps methods and functions for the use of labels.
 *
 * @author MK124
 * @author Marvin Haschker
 */
abstract class Label : Destroyable, Proxyable<Label> {

    /**
     * Gets the internal id of the label.
     */
    abstract val id: Int

    /**
     * Gets the text of the label.
     */
    abstract val text: String

    /**
     * Gets the color of the text.
     */
    abstract val color: Color

    /**
     * Gets the draw distance of the label.
     */
    abstract val drawDistance: Float

    /**
     * Gets the location of the label.
     */
    abstract val location: Location

    /**
     * Gets the attached [Player].
     * If there is none, this property will be null.
     */
    abstract val attachedPlayer: Player?

    /**
     * Gets the attached [Vehicle].
     * If there is none, this property will be null.
     */
    abstract val attachedVehicle: Vehicle?

    /**
     * Attaches the label to the [player].
     * @param player The Player
     * @param x X-Offset
     * @param y Y-Offset
     * @param z Z-Offset
     */
    abstract fun attach(player: Player, x: Float, y: Float, z: Float)

    /**
     * Attaches the label to the [player].
     * @param player The Player
     * @param offset The Location
     */
    abstract fun attach(player: Player, offset: Vector3D)

    /**
     * Attaches the label to the [vehicle].
     * @param vehicle The Vehicle
     * @param x X-Offset
     * @param y Y-Offset
     * @param z Z-Offset
     */
    abstract fun attach(vehicle: Vehicle, x: Float, y: Float, z: Float)

    /**
     * Attaches the label to the [vehicle].
     * @param vehicle The Vehicle
     * @param offset The Location
     */
    abstract fun attach(vehicle: Vehicle, offset: Vector3D)

    /**
     * Updates the [text] and [color] of the label.
     * @param color New Color
     * @param text New Text
     */
    abstract fun update(color: Color, text: String)

    companion object : Collectable<Label>, Findable<Int, Label> {

        /**
         * Invalid ID for a [Label].
         */
        @JvmField
        val INVALID_ID = 0xFFFF

        /**
         * Gets all available [Label]s.
         */
        @JvmStatic
        override fun get(): Collection<Label> = SampObjectManager.get().labels

        /**
         * Gets a [Label] by it's [id].
         */
        @JvmStatic
        override operator fun get(value: Int): Label? = SampObjectManager.get().getLabel(value)

        /**
         * Create a Label with params. If the Creation fails, it will throw a CreationFailedException.
         * @param text The displayed Text.
         * @param color The color of the Label.
         * @param x X-Position.
         * @param y Y-Position.
         * @param z Z-Position.
         * @param worldId Virtual World ID.
         * @param drawDistance The Drawdistance of the Label.
         * @param testLOS If the Label can be seen through objects.
         * @return The created Label.
         */
        @Throws(CreationFailedException::class)
        @JvmStatic
        @JvmOverloads
        fun create(text: String, color: Color = Color.WHITE, x: Float, y: Float, z: Float, worldId:
                    Int = 0, drawDistance: Float = 0f, testLOS: Boolean = false) =
                create(text, color, Location(x, y, z, worldId = worldId), drawDistance, testLOS)

        /**
         * Create a Label with params. If the Creation fails, it will throw a CreationFailedException.
         * @param text The displayed Text.
         * @param color The color of the Label.
         * @param loc Location where the Label should be.
         * @param drawDistance The Drawdistance of the Label.
         * @param testLOS If the Label can be seen through objects.
         * @return The created Label.
         */
        @Throws(CreationFailedException::class)
        @JvmStatic
        @JvmOverloads
        fun create(text: String, color: Color = Color.WHITE, loc: Location = Location(), drawDistance: Float = 50f,
                   testLOS: Boolean = false): Label =
                SampObjectManager.get().createLabel(text, color, loc, drawDistance, testLOS)
    }
}
